extension/main.ts

```diff
@@ -6,14 +6,82 @@
 import { exists, isDirectory, isFile, mkdir, readdir, readFile, rmdir, writeFile } from '@azure-tools/async-io';
 import { Progress, Subscribe } from '@azure-tools/eventing';
 import { CriticalSection, Delay, Exception, Mutex, shallowCopy, SharedLock } from '@azure-tools/tasks';
-import { ChildProcess, spawn, SpawnOptions } from 'child_process';
+import { ChildProcess, spawn, spawnSync, SpawnOptions } from 'child_process';
 import { resolve as npmResolvePackage } from 'npm-package-arg';
 import { homedir, tmpdir } from 'os';
 import * as pacote from 'pacote';
 import { basename, delimiter, dirname, extname, isAbsolute, join, normalize, resolve } from 'path';
 import * as semver from 'semver';
 import { readFileSync } from 'fs';
 
+export async function updatePythonPath(command: Array<string>) {
+  const detect = quoteIfNecessary('import sys; print(sys.hexversion >= 0x03060000)');
+  const path = process.env[getPathVariableName()];
+
+  const pyexe = process.env['AUTOREST_PYTHON_EXE'];
+  if (pyexe) {
+    command[0] = pyexe;
+    return;
+  }
+
+  // detect the python interpreter.
+  const py = await getFullPath('py', path);
+  try {
+    if (py && process.platform === 'win32') {
+      // check if 'py -3' works
+      if (spawnSync(py, ['-3', '-c', detect], { encoding: 'utf8', shell: true }).stdout.toLowerCase().trim() === 'true') {
+        command[0] = py;
+        command.splice(1, 0, '-3');
+        return;
+      }
+    }
+  }
+  catch {
+    // no worries
+  }
+
+  const python3 = await getFullPath('python3', path);
+  try {
+    if (python3) {
+      // check if 'python3' works
+      if (spawnSync(python3, ['-c', detect], { encoding: 'utf8', shell: true }).stdout.toLowerCase().trim() === 'true') {
```

Comment:
Since this appears three times, might be nice to make a simple helper function for the spawnSync and result test so that you only have to pass in Python path and args.

Reply:
Yeah, I started it that way, first, I had too much crap in the way, then I had to fix the parameters because I changed my mind, then I spent two and a half hours chasing a bug in Windows/Node with regards to running cli Windows Store Apps binaries from the path and by the time I was done in the end the function ended up just being a one-liner anyway. 

I left them basically duplicates because the slight variations (in that updating the command array varies, and the spawn itself varies) and ... it's a single line.

Reply:
I'm actually still really irked about the Windows Store glitch. (we can talk about that one!)